generators/example/src/main/scala/GCD.scala

```diff
@@ -0,0 +1,207 @@
+package example
+
+import chisel3._
+import chisel3.util._
+import chisel3.core.{IntParam, Reset}
+import freechips.rocketchip.amba.axi4._
+import freechips.rocketchip.subsystem.BaseSubsystem
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.regmapper.{HasRegMap, RegField}
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util.UIntIsOneOf
+
+// DOC include start: GCD params
+case class GCDParams(
+  address: BigInt = 0x2000,
+  width: Int = 32,
+  useAXI4: Boolean = false,
+  useBlackBox: Boolean = true)
+// DOC include end: GCD params
+
+// DOC include start: GCD key
+case object GCDKey extends Field[Option[GCDParams]](None)
+// DOC include end: GCD key
+
+class GCDIO(w: Int) extends Bundle {
+  val clock = Input(Clock())
+  val reset = Input(Bool())
+  val input_ready = Output(Bool())
+  val input_valid = Input(Bool())
+  val x = Input(UInt(w.W))
+  val y = Input(UInt(w.W))
+  val output_ready = Input(Bool())
+  val output_valid = Output(Bool())
+  val gcd = Output(UInt(w.W))
+  val busy = Output(Bool())
+}
+
+trait GCDTopIO extends Bundle {
+  val gcd_busy = Output(Bool())
+}
+
+// DOC include start: GCD blackbox
+class GCDMMIOBlackBox(w: Int) extends BlackBox(Map("WIDTH" -> IntParam(w))) with HasBlackBoxResource
```

Comment:
Something like this would make the code below a little DRYer:

```suggestion
trait IsGCDImplementation extends BaseModule { val io = IO(new GCDIO(w)) }

class GCDMMIOBlackBox(w: Int) extends BlackBox(Map("WIDTH" -> IntParam(w))) with HasBlackBoxResource with IsGCDImplementation
```

then delete the other `io` definitions.

Reply:
The lack of parameterized traits complicates this, I wasn't able to get this to work with
```
trait HasGCDIO extends BaseModule {
  val w: Int
  val io = IO(new GCDIO(w))
}
class GCDMMIOBlackBox(val w: Int) extends BlackBox(Map("WIDTH" -> IntParam(w))) with HasBlackBoxResource
  with HasGCDIO
{
  addResource("/vsrc/GCDMMIOBlackBox.v")
}
```
I see the error:
```
[deprecated] class example.GCDTL$$anonfun$1$$anon$1 (1 calls): Unable to automatically infer cloneType on class example.GCDTL$$anonfun$1$$anon$1: constructor has parameters ($outer, x$2$1) that are not both immutable and accessible. Either make all parameters immutable and accessible (vals) so cloneType can be inferred, or define a custom cloneType method.
[error] (run-main-0) java.lang.ClassCastException: chisel3.internal.firrtl.Ref cannot be cast to chisel3.internal.firrtl.ModuleIO
[error] java.lang.ClassCastException: chisel3.internal.firrtl.Ref cannot be cast to chisel3.internal.firrtl.ModuleIO
[error]         at chisel3.experimental.BaseModule.$anonfun$getChiselPorts$2(Module.scala:271)
[error]         at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
[error]         at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
[error]         at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
[error]         at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
[error]         at scala.collection.TraversableLike.map(TraversableLike.scala:238)
[error]         at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
[error]         at scala.collection.AbstractTraversable.map(Traversable.scala:108)
[error]         at chisel3.experimental.BaseModule.getChiselPorts(Module.scala:270)
[error]         at chisel3.experimental.DataMirror$.modulePorts(Data.scala:138)
[error]         at chisel3.experimental.DataMirror$.fullModulePorts(Data.scala:147)
[error]         at chisel3.Data.bindingToString(Data.scala:357)
[error]         at chisel3.Bool.toString(Bits.scala:1025)
[error]         at java.lang.String.valueOf(String.java:2994)
```